File: scripts/pool_builtin_fonts.py
"""Deterministic built-in font table pooling; originals remain the source of truth.

Only literal, immutable arrays emitted by fontconvert.py are accepted. Generated
headers live beneath an environment's build directory and retain the original
names as preprocessor aliases (no pointer/reference objects or runtime setup).
No compiler constant-merging flags, dependencies, or SD font changes are needed.
"""
import argparse
import ast
from collections import Counter
import hashlib
import json
from pathlib import Path
import re


# Field widths/signs follow EpdFontData.h. Sizes include C++ structure padding;
# pooled definitions assert the ABI instead of silently trusting this report.
TYPES = {
    "uint8_t": (1, ((8, False),)),
    "int8_t": (1, ((8, True),)),
    "uint16_t": (2, ((16, False),)),
    "EpdGlyph": (16, ((8, False), (8, False), (16, False), (16, True),
                      (16, True), (16, False), (32, False))),
    "EpdFontGroup": (20, ((32, False), (32, False), (32, False), (16, False), (32, False))),
    "EpdUnicodeInterval": (12, ((32, False),) * 3),
    "EpdKernClassEntry": (3, ((16, False), (8, False))),
    "EpdLigaturePair": (8, ((32, False),) * 2),
}
COMMENT = re.compile(r"/\*.*?\*/|//[^\n]*", re.S)
ARRAY = re.compile(r"static\s+const\s+(\w+)\s+(\w+)\s*\[\s*(\d*)\s*\]\s*=\s*(\{.*?\})\s*;", re.S)


def parse_arrays(source):
    """Return validated literal arrays with exact source spans for replacement.

    Reject unknown declarations, expressions, dimensions, types, field counts,
    and out-of-range values. Comments are blanked without changing offsets.
    """
    clean = COMMENT.sub(lambda match: " " * len(match[0]), source)
    arrays = []
    for match in ARRAY.finditer(clean):
        kind, name, extent, body = match.groups()
        if kind not in TYPES:
            raise ValueError(f"Unsupported table type: {kind} {name}")
        residue = re.sub(r"-?(?:0[xX][0-9a-fA-F]+|[0-9]+)|[{},\s]", "", body)
        if residue:
            raise ValueError(f"Nonliteral initializer: {name}")
        try:
            values = ast.literal_eval(body.replace("{", "[").replace("}", "]"))
        except (ValueError, SyntaxError) as exc:
            raise ValueError(f"Unsupported initializer: {name}") from exc
        size, fields = TYPES[kind]
        structured = kind.startswith("Epd")
        if not values or (extent and int(extent) != len(values)):
            raise ValueError(f"Invalid extent: {name}")
        for value in values:
            row = value if structured else [value]
            if not isinstance(row, list) or len(row) != len(fields):
                raise ValueError(f"Invalid field count: {name}")
            for number, (bits, signed) in zip(row, fields):
                low = -(1 << (bits - 1)) if signed else 0
                high = (1 << (bits - int(signed))) - 1
                if type(number) is not int or not low <= number <= high:
                    raise ValueError(f"Invalid field value: {name}")
        frozen = tuple(tuple(row) for row in values) if structured else tuple(values)
        arrays.append({"name": name, "type": kind, "key": (kind, frozen),
                       "start": match.start(), "end": match.end(),
                       "declaration": source[match.start():match.end()],
                       "bytes": len(values) * size})
    remainder = clean
    for array in reversed(arrays):
        remainder = remainder[:array["start"]] + remainder[array["end"]:]
    if re.search(r"\bstatic\s+const\b", remainder):
        raise ValueError("Unsupported static const declaration in font header")
    return arrays


def font_paths(source):
    """Read only the explicit font manifest, preserving both selection branches."""
    manifest = (source / "all.h").read_text()
    reading = re.findall(r"^#include BUILTIN_READING_FONT_HEADER\((\w+)\)$", manifest, re.M)
    ui = re.findall(r"^#include <builtinFonts/(\w+\.h)>$", manifest, re.M)
    if not (reading or ui) or len(reading) + len(ui) != len(re.findall(r"^#include\b", manifest, re.M)):
        raise ValueError("Unsupported built-in font manifest")
    return sorted([Path(name + ".h") for name in reading]
                  + [Path("noemoji") / (name + ".h") for name in reading]
                  + [Path(name) for name in ui])


def pool_symbol(key):
    digest = hashlib.sha256(json.dumps(key, separators=(",", ":")).encode()).hexdigest()
    return "crossink_font_pool_" + digest


def alias_text(name, symbol):
    return f"#include <builtinFonts/pool/{symbol}.h>\n#define {name} {symbol}"


def write_changed(path, text):
    path.parent.mkdir(parents=True, exist_ok=True)
    if not path.exists() or path.read_text() != text:
        path.write_text(text)


def generate(source, output):
    source, output = Path(source), Path(output)
    destination = (output / "builtinFonts").resolve()
    if (destination == source.resolve() or source.resolve() in destination.parents
            or destination in source.resolve().parents):
        raise ValueError("Generated output must be outside source font directory")
    headers = {}
    groups = {}
    # Validate everything before writing output. A parse failure aborts the build.
    for relative in font_paths(source):
        text = (source / relative).read_text()
        arrays = parse_arrays(text)
        headers[relative] = (text, arrays)
        for array in arrays:
            groups.setdefault(array["key"], []).append(array)
    shared = {key: entries for key, entries in groups.items() if len(entries) > 1}
    for key, entries in shared.items():
        first = entries[0]
        symbol = pool_symbol(key)
        declaration = re.sub(r"\b" + re.escape(first["name"]) + r"\b", symbol, first["declaration"], count=1)
        header = ("#pragma once\n#include <EpdFontData.h>\n"
                  f"static_assert(sizeof({first['type']}) == {TYPES[first['type']][0]}, \"Font table ABI changed\");\n"
                  + declaration + "\n")
        write_changed(output / "builtinFonts/pool" / (symbol + ".h"), header)
    for relative, (text, arrays) in headers.items():
        for array in reversed(arrays):
            if array["key"] in shared:
                text = (text[:array["start"]] + alias_text(array["name"], pool_symbol(array["key"]))
                        + text[array["end"]:])
        write_changed(output / "builtinFonts" / relative, text)
    write_changed(output / "builtinFonts/all.h", (source / "all.h").read_text())
    report = {}
    for variant in ("default", "noemoji"):
        selected = [array for path, (_, arrays) in headers.items()
                    if ((Path("noemoji") / path.name not in headers)
                        or (("noemoji" in path.parts) == (variant == "noemoji")))
                    for array in arrays]
        counts = Counter(array["key"] for array in selected)
        savings = Counter()
        for key, count in counts.items():
            savings[key[0]] += (count - 1) * groups[key][0]["bytes"]
        report[variant] = {"array_count": len(selected), "candidate_bytes_saved": sum(savings.values()),
                           "savings_by_type": dict(sorted(savings.items()))}
    write_changed(output / "font-pool-report.json", json.dumps(report, indent=2, sort_keys=True) + "\n")
    return report


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--source", type=Path, default=Path("lib/EpdFont/builtinFonts"))
    parser.add_argument("--output", type=Path, required=True)
    args = parser.parse_args()
    print(json.dumps(generate(args.source, args.output), indent=2, sort_keys=True))

File: lib/EpdFont/scripts/pool-builtin-fonts.rs
//! Deduplicate literal immutable font tables without changing their compiled representation.
use regex::Regex;
use serde_json::{json, Value};
use sha2::{Digest, Sha256};
use std::{
    collections::BTreeMap,
    env,
    error::Error,
    fs,
    path::{Path, PathBuf},
    process::ExitCode,
};
type Result<T> = std::result::Result<T, Box<dyn Error>>;

#[derive(Debug)]
struct Array {
    name: String,
    kind: String,
    key: String,
    start: usize,
    end: usize,
    declaration: String,
    bytes: usize,
}
fn fields(kind: &str) -> Result<(usize, Vec<(u32, bool)>)> {
    Ok(match kind {
        "uint8_t" => (1, vec![(8, false)]),
        "int8_t" => (1, vec![(8, true)]),
        "uint16_t" => (2, vec![(16, false)]),
        "EpdGlyph" => (
            16,
            vec![
                (8, false),
                (8, false),
                (16, false),
                (16, true),
                (16, true),
                (16, false),
                (32, false),
            ],
        ),
        "EpdFontGroup" => (
            20,
            vec![
                (32, false),
                (32, false),
                (32, false),
                (16, false),
                (32, false),
            ],
        ),
        "EpdUnicodeInterval" => (12, vec![(32, false); 3]),
        "EpdKernClassEntry" => (3, vec![(16, false), (8, false)]),
        "EpdLigaturePair" => (8, vec![(32, false); 2]),
        _ => return Err(format!("Unsupported table type: {kind}").into()),
    })
}
fn parse_arrays(source: &str) -> Result<Vec<Array>> {
    let comments = Regex::new(r"(?s)/\*.*?\*/|//[^\n]*")?;
    let clean = comments.replace_all(source, |c: &regex::Captures<'_>| " ".repeat(c[0].len()));
    let pattern =
        Regex::new(r"(?s)static\s+const\s+(\w+)\s+(\w+)\s*\[\s*(\d*)\s*\]\s*=\s*(\{.*?\})\s*;")?;
    let hex = Regex::new(r"-?0[xX][0-9a-fA-F]+")?;
    let trailing = Regex::new(r",\s*]")?;
    let mut result = Vec::new();
    let mut remainder = clean.to_string();
    for cap in pattern.captures_iter(&clean) {
        let whole = cap.get(0).ok_or("Missing array match")?;
        let kind = &cap[1];
        let name = &cap[2];
        let (size, fields) = fields(kind)?;
        let decimal = hex.replace_all(&cap[4], |c: &regex::Captures<'_>| {
            let s = &c[0];
            let negative = s.starts_with('-');
            let value = i64::from_str_radix(&s[if negative { 3 } else { 2 }..], 16);
            match value {
                Ok(n) => (if negative { -n } else { n }).to_string(),
                Err(_) => "INVALID".into(),
            }
        });
        let body = decimal.replace('{', "[").replace('}', "]");
        let body = trailing.replace_all(&body, "]");
        let values: Vec<Value> = serde_json::from_str(&body)
            .map_err(|e| format!("Nonliteral initializer {name}: {e}"))?;
        if values.is_empty() || (!cap[3].is_empty() && cap[3].parse::<usize>()? != values.len()) {
            return Err(format!("Invalid extent: {name}").into());
        }
        for value in &values {
            let row = if kind.starts_with("Epd") {
                value
                    .as_array()
                    .ok_or("Invalid structured initializer")?
                    .as_slice()
            } else {
                std::slice::from_ref(value)
            };
            if row.len() != fields.len() {
                return Err(format!("Invalid field count: {name}").into());
            }
            for (value, (bits, signed)) in row.iter().zip(&fields) {
                let n = value.as_i64().ok_or("Invalid integer field")?;
                let low = if *signed { -(1i64 << (bits - 1)) } else { 0 };
                let high = (1i64 << (bits - u32::from(*signed))) - 1;
                if n < low || n > high {
                    return Err(format!("Invalid field value: {name}").into());
                }
            }
        }
        result.push(Array {
            name: name.into(),
            kind: kind.into(),
            key: serde_json::to_string(&json!([kind, values]))?,
            start: whole.start(),
            end: whole.end(),
            declaration: source[whole.range()].into(),
            bytes: values.len() * size,
        });
    }
    for a in result.iter().rev() {
        remainder.replace_range(a.start..a.end, "");
    }
    if Regex::new(r"\bstatic\s+const\b")?.is_match(&remainder) {
        return Err("Unsupported static const declaration in font header".into());
    }
    Ok(result)
}
fn font_paths(source: &Path) -> Result<Vec<PathBuf>> {
    let text = fs::read_to_string(source.join("all.h"))?;
    let reading = Regex::new(r"^#include BUILTIN_READING_FONT_HEADER\((\w+)\)$")?;
    let ui = Regex::new(r"^#include <builtinFonts/(\w+\.h)>$")?;
    let mut paths = Vec::new();
    for line in text.lines().filter(|line| line.starts_with("#include")) {
        if let Some(c) = reading.captures(line) {
            paths.push(PathBuf::from(format!("{}.h", &c[1])));
            paths.push(PathBuf::from(format!("noemoji/{}.h", &c[1])));
        } else if let Some(c) = ui.captures(line) {
            paths.push(PathBuf::from(&c[1]));
        } else {
            return Err("Unsupported built-in font manifest".into());
        }
    }
    if paths.is_empty() {
        return Err("Empty built-in font manifest".into());
    }
    paths.sort();
    Ok(paths)
}
fn symbol(key: &str) -> String {
    format!("crossink_font_pool_{:x}", Sha256::digest(key.as_bytes()))
}
fn write_changed(path: &Path, text: &str) -> Result<()> {
    if fs::read_to_string(path).ok().as_deref() != Some(text) {
        fs::create_dir_all(path.parent().ok_or("Missing parent directory")?)?;
        fs::write(path, text)?;
    }
    Ok(())
}
fn resolved(path: &Path) -> Result<PathBuf> {
    if path.exists() {
        return Ok(path.canonicalize()?);
    }
    let parent = path
        .parent()
        .filter(|p| !p.as_os_str().is_empty())
        .unwrap_or(Path::new("."));
    Ok(resolved(parent)?.join(path.file_name().ok_or("Invalid output path")?))
}
fn generate(source: &Path, output: &Path) -> Result<Value> {
    let source = source.canonicalize()?;
    let destination = resolved(&output.join("builtinFonts"))?;
    if destination.starts_with(&source) || source.starts_with(&destination) {
        return Err("Generated output must be outside source font directory".into());
    }
    let mut headers = BTreeMap::new();
    for path in font_paths(&source)? {
        let text = fs::read_to_string(source.join(&path))?;
        let arrays = parse_arrays(&text)?;
        headers.insert(path, (text, arrays));
    }
    let mut groups: BTreeMap<&str, Vec<&Array>> = BTreeMap::new();
    for (_, arrays) in headers.values() {
        for a in arrays {
            groups.entry(&a.key).or_default().push(a);
        }
    }
    for (key, arrays) in groups.iter().filter(|(_, a)| a.len() > 1) {
        let a = arrays[0];
        let sym = symbol(key);
        let declaration = a.declaration.replacen(&a.name, &sym, 1);
        write_changed(&destination.join("pool").join(format!("{sym}.h")), &format!("#pragma once\n#include <EpdFontData.h>\nstatic_assert(sizeof({}) == {}, \"Font table ABI changed\");\n{declaration}\n",a.kind,fields(&a.kind)?.0))?;
    }
    for (path, (text, arrays)) in &headers {
        let mut transformed = text.clone();
        for a in arrays.iter().rev() {
            if groups[a.key.as_str()].len() > 1 {
                let sym = symbol(&a.key);
                transformed.replace_range(
                    a.start..a.end,
                    &format!(
                        "#include <builtinFonts/pool/{sym}.h>\n#define {} {sym}",
                        a.name
                    ),
                );
            }
        }
        write_changed(&destination.join(path), &transformed)?;
    }
    write_changed(
        &destination.join("all.h"),
        &fs::read_to_string(source.join("all.h"))?,
    )?;
    let mut report = serde_json::Map::new();
    for variant in ["default", "noemoji"] {
        let mut counts: BTreeMap<&str, usize> = BTreeMap::new();
        let mut count = 0;
        for (path, (_, arrays)) in &headers {
            let name = path.file_name().ok_or("Invalid font path")?;
            if headers.contains_key(&Path::new("noemoji").join(name))
                && path.starts_with("noemoji") != (variant == "noemoji")
            {
                continue;
            }
            for a in arrays {
                *counts.entry(&a.key).or_default() += 1;
                count += 1;
            }
        }
        let mut savings: BTreeMap<&str, usize> = BTreeMap::new();
        for (key, n) in counts {
            let a = groups[key][0];
            *savings.entry(&a.kind).or_default() += (n - 1) * a.bytes;
        }
        report.insert(variant.into(),json!({"array_count":count,"candidate_bytes_saved":savings.values().sum::<usize>(),"savings_by_type":savings}));
    }
    let report = Value::Object(report);
    write_changed(
        &output.join("font-pool-report.json"),
        &(serde_json::to_string_pretty(&report)? + "\n"),
    )?;
    Ok(report)
}
fn run() -> Result<()> {
    let mut source = PathBuf::from("lib/EpdFont/builtinFonts");
    let mut output = None;
    let mut args = env::args_os().skip(1);
    while let Some(arg) = args.next() {
        match arg.to_str() {
            Some("--source") => source = args.next().ok_or("--source requires a directory")?.into(),
            Some("--output") => {
                output = Some(PathBuf::from(
                    args.next().ok_or("--output requires a directory")?,
                ))
            }
            Some("--help" | "-h") => {
                println!("Usage: pool-builtin-fonts [--source DIRECTORY] --output DIRECTORY");
                return Ok(());
            }
            _ => return Err(format!("Unknown argument: {}", arg.to_string_lossy()).into()),
        }
    }
    println!(
        "{}",
        serde_json::to_string_pretty(&generate(&source, &output.ok_or("--output is required")?)?)?
    );
    Ok(())
}
fn main() -> ExitCode {
    match run() {
        Ok(()) => ExitCode::SUCCESS,
        Err(e) => {
            eprintln!("pool-builtin-fonts: {e}");
            ExitCode::FAILURE
        }
    }
}
